File: server/src/main/java/com/bharath/stacked/modules/file/service/FileService.java
package com.bharath.stacked.modules.file.service;

import com.bharath.stacked.modules.file.dto.FileContent;
import com.bharath.stacked.modules.file.dto.FileResponse;
import org.jspecify.annotations.NonNull;
import org.springframework.web.multipart.MultipartFile;

public interface FileService {

    /**
     * Upload a file to temporary storage. Automatically expires based on configured
     * TTL.
     */
    @NonNull
    FileResponse uploadTemporaryFile(@NonNull MultipartFile file, @NonNull String ownerId);

    /**
     * Upload a file directly to permanent storage.
     */
    @NonNull
    FileResponse uploadPermanentFile(@NonNull MultipartFile file, @NonNull String ownerId);

    /**
     * Move an existing temporary file to permanent storage, verifying ownership.
     */
    @NonNull
    FileResponse makePermanent(@NonNull String fileId, @NonNull String ownerId);

    /**
     * Move a temporary file to permanent storage for internal module consumers
     * (e.g. Assessment/Question bank).
     */
    @NonNull
    FileResponse moveToPermanent(@NonNull String fileId);

    /**
     * Retrieve file content for streaming or download (public access for browser
     * media).
     */
    @NonNull
    FileContent getFileContent(@NonNull String fileId);

    /**
     * Retrieve file content for streaming or download.
     */
    @NonNull
    FileContent getFileContent(@NonNull String fileId, @NonNull String authenticatedUserId);

    /**
     * Retrieve file metadata.
     */
    @NonNull
    FileResponse getFileMetadata(@NonNull String fileId, @NonNull String authenticatedUserId);

    /**
     * Delete a file permanently. Only the file owner can delete it.
     */
    void deleteFile(@NonNull String fileId, @NonNull String ownerId);

    /**
     * Clean up expired temporary files from disk and database.
     * 
     * @return count of deleted files
     */
    long cleanupExpiredTemporaryFiles();
}

File: client/src/services/file-service.ts
import API from "@/lib/api-client"
import type { ApiResponse } from "@/types/api"
import type { FileMetadata } from "@/types/file"

export const fileService = {
  /**
   * Upload a file to temporary storage.
   */
  async uploadTemporary(file: File): Promise<FileMetadata> {
    const formData = new FormData()
    formData.append("file", file)

    const response = await API.post<ApiResponse<FileMetadata>>(
      "/files/temporary",
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    )
    return response.data.data
  },

  /**
   * Upload a file directly to permanent storage.
   */
  async uploadPermanent(file: File): Promise<FileMetadata> {
    const formData = new FormData()
    formData.append("file", file)

    const response = await API.post<ApiResponse<FileMetadata>>(
      "/files/permanent",
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    )
    return response.data.data
  },

  /**
   * Move a temporary file to permanent storage.
   */
  async makePermanent(fileId: string): Promise<FileMetadata> {
    const response = await API.post<ApiResponse<FileMetadata>>(
      `/files/${encodeURIComponent(fileId)}/make-permanent`
    )
    return response.data.data
  },

  /**
   * Retrieve metadata for a file.
   */
  async getMetadata(fileId: string): Promise<FileMetadata> {
    const response = await API.get<ApiResponse<FileMetadata>>(
      `/files/${encodeURIComponent(fileId)}`
    )
    return response.data.data
  },

  /**
   * Download or retrieve raw file content as a Blob.
   */
  async getFileBlob(fileId: string): Promise<Blob> {
    const response = await API.get<Blob>(
      `/files/${encodeURIComponent(fileId)}/content`,
      {
        responseType: "blob",
      }
    )
    return response.data
  },

  /**
   * Permanently delete a file. Caller must be the file owner.
   */
  async deleteFile(fileId: string): Promise<void> {
    await API.delete<ApiResponse<void>>(`/files/${encodeURIComponent(fileId)}`)
  },
}
